### PR TITLE
Implement filetree.

### PR DESCRIPTION
**Pull request by kadams54**

### CODE REVIEW

**Comment by kadams54** [comment]
## QA

1. Pull up the entity details page for a charm.
2. Verify that directory contents are collapsed in the file listing.
3. Verify that clicking on the directory expands the listing.
4. Repeat for a bundle.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3768/
Test PASSed.

**Comment by hatched** [comment]
QA Issue: 

`/?store=trusty%2Fredis`
Click on the text for the items with the + and it adds another + instead of opening.

**Comment by hatched** [comment]
QA Issue: 

The + and - in the box appears to be slightly off centre.

File: jujugui/static/gui/src/app/components/entity-files/entity-files.js [hunk 1]
```diff
@@ -64,17 +118,53 @@ YUI.add('entity-files', function() {
...
+          );
+        } else {
+          var childItems = [];
+          Object.keys(children).forEach(function(child) {
+            childItems.push(buildList(`${path}/${child}`, children[child]));
+          });
+          return (
+            <li key={path}
+              className="entity-files__directory collapsed"
+              onClick={clickHandler}>
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/entity-files/entity-files.js]: Clickable items need `tabIndex="0" role="button"` for keyboard navigation and accessibility.

**Comment by anthonydillon** [comment]
QA issues:
- Missing connecting lines
- Remove the bullet styles
- Clicking the directory or file name it adds a plus icon
- Directories should start with forward slash

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3797/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3805/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3817/
Test PASSed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3818/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3819/
Test PASSed.

**Comment by hatched** [comment]
QA Issue: 
You'll see that the last file in the list is not an elbow like it is in jc.c 

screenshot from 2015-12-07 10 00 43
vs
screenshot from 2015-12-07 10 01 17

File: jujugui/static/gui/src/app/components/entity-files/entity-files.js [hunk 2]
```diff
@@ -64,17 +118,55 @@ YUI.add('entity-files', function() {
...
+                {childItems}
+              </ul>
+            </li>
+          );
+        }
+      }
+
+      // Need to handle the root node separate from the recursive logic.
+      var markup = [];
+      Object.keys(filetree).forEach(function(file) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-files/entity-files.js]: If you use an arrow fn here I believe you can get rid of that `bind` on 122 which is making react complain.

Comment (kadams54) [inline jujugui/static/gui/src/app/components/entity-files/entity-files.js]: Gave it a whirl and it didn't work - `Window` was assigned to `this` by the time it got into the recursive function. I think the problem here isn't the `forEach` but rather the recursion.

**Comment by kadams54** [comment]
I did some investigating into the "elbow" issue. jujucharms.com has the "elbow" on files only. If you view a directory that has another directory as it's last child, you can see the border hanging down:

jujucharms.com directory

The solution jujucharms.com uses the make an elbow also prevents us from dealing with overflow, so you end up with directory listings that look like this:

jujucharms.com wrap

Personally I'd rather have ellipsis on long file names and consistent non-elbows for both files and directories:

truncated with no elbows

But I can do elbows and wrapping if we want 100% compatibility with jujucharms.com.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3837/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3838/
Test FAILed.

**Comment by hatched** [comment]
:+1: QA :ok: Thanks for stickin on this.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3839/
Test PASSed.

**Comment by kadams54** [comment]
Thanks all! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3840/
Test PASSed.

**Pull request merged**